crates/bevy_physics/src/d3/ray.rs

```diff
@@ -0,0 +1,73 @@
+use bevy_render::camera::Camera;
+use bevy_transform::components::GlobalTransform;
+use bevy_window::Window;
+use glam::{Vec2, Vec3};
+
+pub struct Ray {
+    origin: Vec3,
+    direction: Vec3,
+}
+
+impl Ray {
+    pub fn new(origin: Vec3, direction: Vec3) -> Self {
+        direction.normalize();
+        Self { origin, direction }
+    }
+
+    pub fn from_window(
+        window: &Window,
+        camera: &Camera,
+        camera_transform: &GlobalTransform,
+    ) -> Self {
+        Self::from_mouse_position(
+            &window.cursor_position().unwrap(),
+            window,
+            camera,
+            camera_transform,
+        )
+    }
+
+    pub fn from_mouse_position(
+        mouse_position: &Vec2,
+        window: &Window,
+        camera: &Camera,
+        camera_transform: &GlobalTransform,
+    ) -> Self {
+        if window.id() != camera.window {
+            panic!("Generating Ray from Camera with wrong Window");
```

Comment:
Correct me if i'm wrong, this can terminate the program?

Reply:
It can, you're right. This is by design. Passing a non-matching window and camera to from_mouse_position(..) would result in undefined behavior and should be checked before calling the method. Do you think a Log Error would be more sensible? I didn't familiarize myself very much with the Diagnostics Plugin, so maybe that could help.

Reply:
TBH I don't know the right answer, I am learning here 👍